src/libs/workspace-hooks.ts

```diff
@@ -0,0 +1,40 @@
+import { useEffect } from 'react';
+import { Ajax } from 'src/libs/ajax';
+import { useLoadedData } from 'src/libs/ajax/loaded-data/useLoadedData';
+import { useCancellation } from 'src/libs/react-utils';
+import { exhaustiveGuard } from 'src/libs/type-utils/type-helpers';
+import { WorkspaceWrapper } from 'src/libs/workspace-utils';
+
+export const useWorkspaceById = (workspaceId: string, fields?: string[]) => {
+  const [workspace, loadWorkspace] = useLoadedData<WorkspaceWrapper>();
+
+  const signal = useCancellation();
+  useEffect(
+    () => {
+      loadWorkspace(() => Ajax(signal).Workspaces.getById(workspaceId, fields));
+    },
+    // Since fields is an array and will likely be defined inline (recreated on each render),
+    // we can't use it directly in the dependencies list. Instead, using a joined list of the
+    // fields provides the behavior we want, rerunning the effect when the fields change.
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+    [loadWorkspace, signal, workspaceId, (fields || []).join(',')]
+  );
+
+  const { status } = workspace;
+  switch (status) {
+    case 'None':
+      return { workspace: null, status: 'None' as const };
+    case 'Loading':
+      // useLoadedData keeps the previous state while loading new data.
+      // When we're loading a different workspace, we don't want to return the previous workspace.
+      return { workspace: null, status: 'Loading' as const };
+    case 'Ready':
+      const { state: readyState, ...readyRest } = workspace;
+      return { ...readyRest, workspace: readyState };
+    case 'Error':
+      const { state: errorState, ...errorRest } = workspace;
+      return { ...errorRest, workspace: errorState };
```

Comment:
Is there a type for what this returns? I have trouble understanding how `…errorRest` translates to `status` (to have the same return shape as in line 30). And why is `errorState` the workspace object?

Reply:
This is renaming the `state` field from the LoadedData type to `workspace`.

Reply:
Found a cleaner way to do this in https://github.com/DataBiosphere/terra-ui/pull/4082/commits/c39896c47700086d3de4d279c219d1e262f6d166.

Reply:
`LoadedState` and `useLoadedData` use a generically named `state` field.
https://github.com/DataBiosphere/terra-ui/blob/1e3201c6f28f248b9da96531f73aa424868e78ec/src/libs/type-utils/LoadedState.ts#L5-L19

For hooks wrapping that, I prefer to use a more descriptively named field for the loaded data (`workspace` for `useWorkspace` in this case). But maybe I should just embrace the `state` field instead of jumping through hoops to rename it.

Reply:
Yeah, I don't know if the renaming is worth it… I'm just not familiar with `LoadedState` (I tried using it once and found it more confusing than handling the state loading myself), so it looked really odd to me. The renaming makes it more understandable what is going on, but it still doesn't provide the type clarity that I would like.

Do you find `LoadedState` useful? I'm wondering how much we want to push on it as a pattern.

Reply:
I think the general type pattern it uses is useful, but I'm not thrilled with it. I end up renaming the generic `state` field every time I use it. Most of the time, I want to immediately load something on mount, which requires adding another `useEffect` hook (which means I have to handle the complexity of the dependencies list), and the `NoneState` is unwieldy: since it doesn't include a `state` field, I often have to add a special case to handle NoneState, which usually isn't even relevant since I'm immediately loading the data on mount.

Reply:
Removed `useLoadedData` from this PR in https://github.com/DataBiosphere/terra-ui/pull/4082/commits/0d5129466352e9c3db25bfe80f61d6bb7b7006e2.